bip-encrypted-backup.md

```diff
@@ -0,0 +1,268 @@
+interactive workflows.
+
+## Specification
+
+Note: in the followings sections, the operator ⊕ refers to the bitwise XOR.
+
+### Secret generation
+
+* Let `p_1`, `p_2`, .., `p_n` be the public keys in the descriptor/wallet policy, in
+increasing lexicographical order.  
+* let `secret` = sha256("BEB_DECRYPTION_SECRET" ‖ `p1` ‖ `p2` ‖ ... ‖ `pn`)  
+* let `si` = sha256("BEB_INDIVIDUAL_SECRET" ‖ `pi`)  
+* let `individual_secret_i` = `secret` ⊕ `si`
+
+### AES-GCM Encryption
+
+* let `nonce` = random()  
+* let `ciphertext` = aes_gcm_256_encrypt(`payload`, `secret`, `nonce`)
+
+### AES-GCM Decryption
+
+In order to decrypt the payload of a backup, the owner of a certain public key p
+computes:
+
+* let `si` = sha256("BEB_INDIVIDUAL_SECRET" ‖ `p`)  
+* for each `individual_secret_i` generate `reconstructed_secret_i` =
+`individual_secret_i` ⊕ `si`  
+* for each `reconstructed_secret_i` process `payload` =
+aes_gcm_256_decrypt(`ciphertext`, `secret`, `nonce`)
+
+Decryption will succeed if and only if **p** was one of the keys in the
+descriptor/wallet policy.
+
+### Encoding
+
+The encrypted backup must be encoded as follows:
+
+`MAGIC` `VERSION` `DERIVATION_PATHS` `INDIVIDUAL_SECRETS` `CONTENT` `ENCRYPTION`
+`ENCRYPTED_PAYLOAD`
+
+#### Magic
+
+`MAGIC`: 3 bytes which are ASCII/UTF-8 representation of **BEB** (`0x42, 0x45,
+0x42`).
+
+#### Version
+
+`VERSION`: 1 byte unsigned integer representing the format version. The current
+specification defines version `0x01`.
+
+#### Derivation Paths
+
+ Note: the derivation-path vector should not contain duplicates.  
+ Derivation paths are optional; they can be useful to simplify the recovery process
+if one has used a non-common derivation path to derive his key.
+
+`DERIVATION_PATH` follows this format:
+
+`COUNT`  
+`CHILD_COUNT` `CHILD` `...` `CHILD`  
+`...`  
+`CHILD_COUNT` `CHILD` `...` `CHILD`  
+
+`COUNT`: 1-byte unsigned integer (0–255) indicating how many derivation paths are
+included.  
+`CHILD_COUNT`: 1-byte unsigned integer (1–255) indicating how many children are in
+the current path.  
+`CHILD`: 4-byte big-endian unsigned integer representing a child index per BIP-32.
+
+#### Individual Secrets
+
+At least one individual secret must be supplied.
+
+The `INDIVIDUAL_SECRETS` section follows this format:
+
+`COUNT`  
+`INDIVIDUAL_SECRET`  
+`INDIVIDUAL_SECRET`
+
+`COUNT`: 1-byte unsigned integer (1–255) indicating how many secrets are included.  
+`INDIVIDUAL_SECRET`: 32-byte serialization of the derived individual secret.
+
+#### Content
+
+`CONTENT`: 1-byte unsigned integer identifying what has been encrypted.
+
+| Value  | Definition                             |
+|:-------|:---------------------------------------|
+| 0x00   | Undefined                              |
+| 0x01   | BIP-0380 Descriptor (string)           |
+| 0x02   | BIP-0388 Wallet policy (string)        |
+| 0x03   | BIP-0329 Labels (JSONL)                |
+| 0x04   | Wallet backup (JSON)                   |
```

Comment:
Maybe point out that unlisted types should still be decrypted and that new types may be added to the BIP.

That said, a more generic approach would be to allow a bit more space for the type. E.g. a ~4 digit number that must represent a BIP number. Or a short string like "BIP-0329". Both avoid the need to maintain a central register.

Also, if you allow multiple pieces of content in a single encrypted blob, it's useful to have a padding type.